plugin-developer-learning-pathway/07-Custom-Database-Tables/02-Interacting-with-custom-tables.md

```diff
@@ -0,0 +1,372 @@
+
+Consider this example where user input is inserted directly into a query without validation:
+
+```php
+global $wpdb;
+$name = $_GET['name'];
+
+$results = $wpdb->get_results( "SELECT * FROM {$wpdb->prefix}submissions WHERE name = '$name'" );
+```
+
+In this example, an attacker could supply a value like `John'; DROP TABLE wp_submissions;--`, which would delete the entire custom table! Let's see how we can prevent this from happening.
+
+### Using Prepared Statements
+
+To protect against SQL injections, always use **prepared statements**. Prepared statements separate SQL logic from data, ensuring that user inputs are properly sanitized and escaped before being inserted into the query. In WordPress, the [`$wpdb->prepare()`](https://developer.wordpress.org/reference/classes/wpdb/prepare/) method is provided for this purpose. Its used similarly to PHP's [`sprintf()`](https://www.php.net/sprintf) and [`vsprintf()`](https://www.php.net/vsprintf) functions where you use placeholders in your SQL query and pass an array of the corresponding values.
+
+Here’s how you can rewrite the previous query to safely handle user input:
+
+```php
+global $wpdb;
+$name = $_GET['name'];
+
+$prepared_query = $wpdb->prepare( 
+    "SELECT * FROM {$wpdb->prefix}submissions WHERE name = %s", 
+    $name 
+);
+
+$results = $wpdb->get_results( $prepared_query );
+```
+
+In this example:
+- `%s` is a placeholder for a string.
+- The `prepare()` method ensures that the `$name` value is properly escaped before being included in the SQL query.
+
+Notice that we do not keep the quotes around the `%s` placeholder value. By using the appropriate placeholder, the `prepare()` method automatically handles that for us. The following placeholders per data type are:
+
+- `%d` (integer)
+- `%f` (float)
+- `%s` (string)
+- `%i` (identifier, e.g. table/field names, introduced in WordPress 6.2.0)
+
+### Validating Data
+
+In addition to properly escaping dynamic values used in SQL queries, it's also important to validate the data. Even honest users can enter incorrect data, so it's important for your plugin to check the data before using it. This improves data integrity and it also gives your plugin a chance to alert users of what's expected if certain criteria isn't met.
+
+Some common [data validation techniques](https://developer.wordpress.org/apis/security/data-validation/) are:
+
+- **Safelist** – Accept data only if it matches a known, trusted value. (eg. selecting a color option)
+- **Blocklist** – Reject data if it matches a known, untrusted value. (eg. rejecting profanity)
+- **Format Detection** – Test that the data matches an accepted format. (eg. formatting phone numbers)
+
+When comparing untrusted data against the safelist, it’s important to **use strict type checking**. Otherwise, an attacker could craft input in a way that will pass validation against a safelist, but still have a malicious effect.
```

Comment:
There is this that kind of glosses over it: https://www.php.net/manual/en/language.operators.comparison.php
Otherwise, we'd need to just link to PHP's type system documentation it seems because it's actually quite a broad topic as far as type juggling, type declarations, etc. https://www.php.net/manual/en/language.types.type-system.php 🤔